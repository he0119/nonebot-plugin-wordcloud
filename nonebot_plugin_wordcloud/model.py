from datetime import time
from typing import Optional

from nonebot_plugin_orm import Model
from nonebot_plugin_saa import PlatformTarget
from sqlalchemy import JSON
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column


class Schedule(Model):
    """定时发送"""

    id: Mapped[int] = mapped_column(primary_key=True)
    target: Mapped[dict] = mapped_column(JSON().with_variant(JSONB, "postgresql"))
    time: Mapped[Optional["time"]]
    """ UTC 时间 """

    @property
    def saa_target(self) -> PlatformTarget:
        return PlatformTarget.deserialize(self.target)
