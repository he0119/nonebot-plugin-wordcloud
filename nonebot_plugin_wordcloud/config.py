from datetime import datetime, time
from pathlib import Path
from typing import Any, Dict, List, Optional, Set, Union
from zoneinfo import ZoneInfo

from nonebot import get_driver
from nonebot_plugin_localstore import get_data_dir
from pydantic import BaseModel, Extra, root_validator

DATA_DIR = get_data_dir("nonebot_plugin_wordcloud")


class Config(BaseModel, extra=Extra.ignore):
    wordcloud_width: int = 1920
    wordcloud_height: int = 1200
    wordcloud_background_color: str = "black"
    wordcloud_colormap: Union[str, List[str]] = "viridis"
    wordcloud_font_path: str
    wordcloud_stopwords_path: Optional[Path]
    wordcloud_userdict_path: Optional[Path]
    wordcloud_timezone: Optional[str]
    wordcloud_default_schedule_time: time
    """ 默认定时发送时间

    如果群内不单独设置则使用这个值，默认为晚上 10 点，时区为设定的时区
    """
    wordcloud_options: Dict[str, Any] = {}
    wordcloud_exclude_user_ids: Set[str] = set()
    """排除的用户 ID 列表（全局，不区分平台）"""

    @root_validator(pre=True, allow_reuse=True)
    def set_default_values(cls, values):
        if not values.get("wordcloud_font_path"):
            values["wordcloud_font_path"] = str(
                Path(__file__).parent / "SourceHanSans.otf"
            )

        default_schedule_time = (
            time.fromisoformat(wordcloud_default_schedule_time)
            if (
                wordcloud_default_schedule_time := values.get(
                    "wordcloud_default_schedule_time"
                )
            )
            else time(22, 0, 0)
        )

        default_schedule_time = (
            default_schedule_time.replace(tzinfo=ZoneInfo(wordcloud_timezone))
            if (wordcloud_timezone := values.get("wordcloud_timezone"))
            else default_schedule_time.replace(
                tzinfo=datetime.now().astimezone().tzinfo
            )
        )
        values["wordcloud_default_schedule_time"] = default_schedule_time
        return values

    def get_mask_path(self, key: Optional[str] = None) -> Path:
        """获取 mask 文件路径"""
        if key is None:
            return DATA_DIR / "mask.png"
        return DATA_DIR / f"mask-{key}.png"


global_config = get_driver().config
plugin_config = Config.parse_obj(global_config)
