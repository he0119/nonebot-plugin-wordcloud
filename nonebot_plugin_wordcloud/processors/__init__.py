from .base import Processor as Processor
