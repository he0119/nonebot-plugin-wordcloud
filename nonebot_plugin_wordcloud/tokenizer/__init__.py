from .base import Tokenizer as Tokenizer
