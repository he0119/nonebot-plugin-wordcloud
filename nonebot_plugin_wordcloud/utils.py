from datetime import datetime, time
from typing import Optional

from nonebot_plugin_apscheduler import scheduler

from .config import plugin_config

try:
    from zoneinfo import ZoneInfo
except ImportError:
    from backports.zoneinfo import ZoneInfo  # type: ignore


def get_datetime_now_with_timezone() -> datetime:
    """获取当前时间，并包含时区信息"""
    if plugin_config.wordcloud_timezone:
        return datetime.now(ZoneInfo(plugin_config.wordcloud_timezone))
    else:
        return datetime.now().astimezone()


def get_datetime_fromisoformat_with_timezone(date_string: str) -> datetime:
    """从 iso8601 格式字符串中获取时间，并包含时区信息"""
    if plugin_config.wordcloud_timezone:
        raw = datetime.fromisoformat(date_string)
        if raw.tzinfo:
            return raw.astimezone(ZoneInfo(plugin_config.wordcloud_timezone))
        else:
            return raw.replace(tzinfo=ZoneInfo(plugin_config.wordcloud_timezone))
    else:
        return datetime.fromisoformat(date_string).astimezone()


def time_astimezone(time: time, tz: Optional[ZoneInfo] = None) -> time:
    """将 time 对象转换为指定时区的 time 对象

    如果 tz 为 None，则转换为本地时区
    """
    local_time = datetime.combine(datetime.today(), time)
    return local_time.astimezone(tz).timetz()


def get_time_fromisoformat_with_timezone(time_string: str) -> time:
    """从 iso8601 格式字符串中获取时间，并包含时区信息"""
    if plugin_config.wordcloud_timezone:
        raw = time.fromisoformat(time_string)
        if raw.tzinfo:
            return time_astimezone(raw, ZoneInfo(plugin_config.wordcloud_timezone))
        else:
            return raw.replace(tzinfo=ZoneInfo(plugin_config.wordcloud_timezone))
    else:
        return time_astimezone(time.fromisoformat(time_string))


def get_time_with_scheduler_timezone(time: time) -> time:
    """获取转换到 APScheduler 时区的时间"""
    return time_astimezone(time, scheduler.timezone)
