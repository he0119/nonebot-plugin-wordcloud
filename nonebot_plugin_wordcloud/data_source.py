import asyncio
import concurrent.futures
import contextlib
from collections.abc import Iterable
from functools import partial
from io import BytesIO
from random import choice
from typing import Optional

import numpy as np
from nonebot.utils import resolve_dot_notation
from PIL import Image
from wordcloud import WordCloud

from .config import global_config, plugin_config
from .processors import Processor

MSGS_PROCESSORS: list[Processor] = [
    resolve_dot_notation(
        processor,
        "Processor",
        "nonebot_plugin_wordcloud.processors.",
    )()
    for processor in plugin_config.worodcloud_msgs_processors
]


def get_mask(key: str):
    """获取 mask"""
    mask_path = plugin_config.get_mask_path(key)
    if mask_path.exists():
        return np.array(Image.open(mask_path))
    # 如果指定 mask 文件不存在，则尝试默认 mask
    default_mask_path = plugin_config.get_mask_path()
    if default_mask_path.exists():
        return np.array(Image.open(default_mask_path))


def get_frequencies(words: Iterable[str]) -> dict[str, float]:
    """获取词频"""
    stopwords = set()
    if plugin_config.wordcloud_stopwords_path:
        with open(plugin_config.wordcloud_stopwords_path, encoding="utf8") as f:
            stopwords = set(f.read().splitlines())

    frequencies: dict[str, float] = {}
    for word in words:
        if word in stopwords:
            continue
        frequencies[word] = frequencies.get(word, 0) + 1
    return frequencies


def analyse_messages(msgs: Iterable[str]) -> dict[str, float]:
    """分析消息，获取词频"""
    # 过滤掉命令
    command_start = tuple(i for i in global_config.command_start if i)
    msgs = (m for m in msgs if not m.startswith(command_start))
    # 处理消息
    for processor in MSGS_PROCESSORS:
        msgs = processor.process_msgs(msgs)
        if isinstance(msgs, dict):
            return msgs
    # 统计消息频率
    frequencies = get_frequencies(msgs)
    return frequencies


def _get_wordcloud(msgs: list[str], mask_key: str) -> Optional[bytes]:
    frequencies = analyse_messages(msgs)
    # 词云参数
    wordcloud_options = {}
    wordcloud_options.update(plugin_config.wordcloud_options)
    wordcloud_options.setdefault("font_path", str(plugin_config.wordcloud_font_path))
    wordcloud_options.setdefault("width", plugin_config.wordcloud_width)
    wordcloud_options.setdefault("height", plugin_config.wordcloud_height)
    wordcloud_options.setdefault(
        "background_color", plugin_config.wordcloud_background_color
    )
    # 如果 colormap 是列表，则随机选择一个
    colormap = (
        plugin_config.wordcloud_colormap
        if isinstance(plugin_config.wordcloud_colormap, str)
        else choice(plugin_config.wordcloud_colormap)
    )
    wordcloud_options.setdefault("colormap", colormap)
    wordcloud_options.setdefault("mask", get_mask(mask_key))
    with contextlib.suppress(ValueError):
        wordcloud = WordCloud(**wordcloud_options)
        image = wordcloud.generate_from_frequencies(frequencies).to_image()
        image_bytes = BytesIO()
        image.save(image_bytes, format="PNG")
        return image_bytes.getvalue()


async def get_wordcloud(msgs: list[str], mask_key: str) -> Optional[bytes]:
    loop = asyncio.get_running_loop()
    pfunc = partial(_get_wordcloud, msgs, mask_key)
    # 虽然不知道具体是哪里泄漏了，但是通过每次关闭线程池可以避免这个问题
    # https://github.com/he0119/nonebot-plugin-wordcloud/issues/99
    with concurrent.futures.ThreadPoolExecutor() as pool:
        return await loop.run_in_executor(pool, pfunc)
