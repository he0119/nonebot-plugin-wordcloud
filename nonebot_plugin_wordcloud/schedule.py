from datetime import time
from typing import Dict, Optional, Tuple
from zoneinfo import ZoneInfo

import nonebot_plugin_saa as saa
from apscheduler.job import Job
from nonebot.log import logger
from nonebot_plugin_apscheduler import scheduler
from nonebot_plugin_cesaa import get_messages_plain_text
from nonebot_plugin_orm import get_session
from sqlalchemy import JSON, Select, cast, select
from sqlalchemy.ext.asyncio import AsyncSession

from .config import plugin_config
from .data_source import get_wordcloud
from .model import Schedule
from .utils import (
    get_datetime_now_with_timezone,
    get_mask_key,
    get_time_with_scheduler_timezone,
    time_astimezone,
)

saa.enable_auto_select_bot()


class Scheduler:
    def __init__(self):
        # 默认定时任务的 key 为 default
        # 其他则为 ISO 8601 格式的时间字符串
        self.schedules: Dict[str, Job] = {}

        # 转换到 APScheduler 的时区
        scheduler_time = get_time_with_scheduler_timezone(
            plugin_config.wordcloud_default_schedule_time
        )
        # 添加默认定时任务
        self.schedules["default"] = scheduler.add_job(
            self.run_task,
            "cron",
            hour=scheduler_time.hour,
            minute=scheduler_time.minute,
            second=scheduler_time.second,
        )

    async def update(self):
        """更新定时任务"""
        async with get_session() as session:
            statement = (
                select(Schedule.time)
                .group_by(Schedule.time)
                .where(Schedule.time != None)
            )
            schedule_times = await session.scalars(statement)
            for schedule_time in schedule_times:
                assert schedule_time is not None
                time_str = schedule_time.isoformat()
                if time_str not in self.schedules:
                    # 转换到 APScheduler 的时区，因为数据库中的时间是 UTC 时间
                    scheduler_time = get_time_with_scheduler_timezone(
                        schedule_time.replace(tzinfo=ZoneInfo("UTC"))
                    )
                    self.schedules[time_str] = scheduler.add_job(
                        self.run_task,
                        "cron",
                        hour=scheduler_time.hour,
                        minute=scheduler_time.minute,
                        second=scheduler_time.second,
                        args=(schedule_time,),
                    )
                    logger.debug(f"已添加每日词云定时发送任务，发送时间：{time_str} UTC")

    async def run_task(self, time: Optional[time] = None):
        """执行定时任务

        时间为 UTC 时间，并且没有时区信息
        如果没有传入时间，则执行默认定时任务
        """
        async with get_session() as session:
            statement = select(Schedule).where(Schedule.time == time)
            results = await session.scalars(statement)
            schedules = results.all()
            # 如果该时间没有需要执行的定时任务，且不是默认任务则从任务列表中删除该任务
            if time and not schedules:
                self.schedules.pop(time.isoformat()).remove()
                return
            logger.info(f"开始发送每日词云，时间为 {time or '默认时间'}")
            for schedule in schedules:
                target = schedule.saa_target
                dt = get_datetime_now_with_timezone()
                start = dt.replace(hour=0, minute=0, second=0, microsecond=0)
                stop = dt
                messages = await get_messages_plain_text(
                    target=target,
                    types=["message"],
                    time_start=start,
                    time_stop=stop,
                    exclude_id1s=plugin_config.wordcloud_exclude_user_ids,
                )
                mask_key = get_mask_key(target)
                if not (image := await get_wordcloud(messages, mask_key)):
                    await saa.Text("今天没有足够的数据生成词云").send_to(target)
                    continue

                await saa.Image(image).send_to(target)

    async def get_schedule(self, target: saa.PlatformTarget) -> Optional[time]:
        """获取定时任务时间"""
        async with get_session() as session:
            statement = self.select_target_statement(target, session)
            results = await session.scalars(statement)
            if schedule := results.one_or_none():
                if schedule.time:
                    # 将时间转换为本地时间
                    return time_astimezone(
                        schedule.time.replace(tzinfo=ZoneInfo("UTC"))
                    )
                else:
                    return plugin_config.wordcloud_default_schedule_time

    async def add_schedule(
        self, target: saa.PlatformTarget, *, time: Optional[time] = None
    ):
        """添加定时任务

        时间需要带时区信息
        """
        # 将时间转换为 UTC 时间
        if time:
            time = time_astimezone(time, ZoneInfo("UTC"))

        async with get_session() as session:
            statement = self.select_target_statement(target, session)
            results = await session.scalars(statement)
            if schedule := results.one_or_none():
                schedule.time = time
            else:
                schedule = Schedule(time=time, target=target.dict())
                session.add(schedule)
            await session.commit()
        await self.update()

    async def remove_schedule(self, target: saa.PlatformTarget):
        """删除定时任务"""
        async with get_session() as session:
            statement = self.select_target_statement(target, session)
            results = await session.scalars(statement)
            if schedule := results.one_or_none():
                await session.delete(schedule)
                await session.commit()

    @staticmethod
    def select_target_statement(
        target: saa.PlatformTarget, session: AsyncSession
    ) -> Select[Tuple[Schedule]]:
        """获取查询目标的语句

        MySQL 需要手动将 JSON 类型的字段转换为 JSON 类型
        """
        engine = session.get_bind()
        if engine.dialect.name == "mysql":
            return select(Schedule).where(Schedule.target == cast(target.dict(), JSON))
        return select(Schedule).where(Schedule.target == target.dict())


schedule_service = Scheduler()
