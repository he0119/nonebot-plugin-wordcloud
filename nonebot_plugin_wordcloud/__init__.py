""" 词云
"""
from nonebot import require

require("nonebot_plugin_apscheduler")
require("nonebot_plugin_alconna")
require("nonebot_plugin_cesaa")

import re
from datetime import datetime, timedelta
from io import BytesIO
from typing import Optional, Union

import nonebot_plugin_alconna as alc
import nonebot_plugin_saa as saa
from arclet.alconna import ArparmaBehavior
from arclet.alconna.arparma import Arparma
from nonebot import get_driver
from nonebot.adapters import Bot, Event, Message
from nonebot.params import Arg, Depends
from nonebot.permission import SUPERUSER
from nonebot.plugin import PluginMetadata, inherit_supported_adapters
from nonebot.typing import T_State
from nonebot_plugin_alconna import (
    Alconna,
    AlconnaMatch,
    AlconnaMatcher,
    AlconnaQuery,
    Args,
    Match,
    Option,
    Query,
    image_fetch,
    on_alconna,
    store_true,
)
from nonebot_plugin_cesaa import get_messages_plain_text
from nonebot_plugin_session import Session, SessionIdType, extract_session
from PIL import Image

from . import migrations
from .config import Config, plugin_config
from .data_source import get_wordcloud
from .schedule import schedule_service
from .utils import (
    admin_permission,
    ensure_group,
    get_datetime_fromisoformat_with_timezone,
    get_datetime_now_with_timezone,
    get_mask_key,
    get_time_fromisoformat_with_timezone,
)

get_driver().on_startup(schedule_service.update)

__plugin_meta__ = PluginMetadata(
    name="词云",
    description="利用群消息生成词云",
    usage="""\
- 通过快捷命令，以获取常见时间段内的词云
格式：/<时间段>词云
时间段关键词有：今日，昨日，本周，上周，本月，上月，年度
示例：/今日词云，/昨日词云

- 提供日期与时间，以获取指定时间段内的词云（支持 ISO8601 格式的日期与时间，如 2022-02-22T22:22:22）
格式：/历史词云 [日期或时间段]
示例：/历史词云
/历史词云 2022-01-01
/历史词云 2022-01-01~2022-02-22
/历史词云 2022-02-22T11:11:11~2022-02-22T22:22:22

- 在上方所给的命令格式基础上，还可以添加前缀“我的”，以获取自己的词云
格式：/我的<基本命令格式>
示例：/我的今日词云
/我的昨日词云

- 设置自定义词云形状
格式：/设置词云形状
/设置词云形状

- 设置默认词云形状（仅超级用户）
格式：/设置词云默认形状
/删除词云默认形状

- 设置定时发送每日词云
格式：/词云每日定时发送状态
/开启词云每日定时发送
/开启词云每日定时发送 23:59
/关闭词云每日定时发送""",
    homepage="https://github.com/he0119/nonebot-plugin-wordcloud",
    type="application",
    supported_adapters=inherit_supported_adapters(
        "nonebot_plugin_chatrecorder", "nonebot_plugin_saa", "nonebot_plugin_alconna"
    ),
    config=Config,
    extra={"orm_version_location": migrations},
)


class SameTime(ArparmaBehavior):
    def operate(self, interface: Arparma):
        type = interface.query("type")
        time = interface.query("time")
        if type is None and time:
            interface.behave_fail()


wordcloud_cmd = on_alconna(
    Alconna(
        "词云",
        Option("--my", default=False, action=store_true),
        Args["type?", ["今日", "昨日", "本周", "上周", "本月", "上月", "年度", "历史"]]["time?", str],
        behaviors=[SameTime()],
    ),
    use_cmd_start=True,
)
wordcloud_cmd.shortcut(
    r"我的(?P<type>.+)词云",
    {
        "prefix": True,
        "command": "词云",
        "args": ["--my", "{type}"],
    },
)
wordcloud_cmd.shortcut(
    r"(?P<type>.+)词云",
    {
        "prefix": True,
        "command": "词云",
        "args": ["{type}"],
    },
)


def parse_datetime(key: str):
    """解析数字，并将结果存入 state 中"""

    async def _key_parser(
        matcher: AlconnaMatcher,
        state: T_State,
        input: Union[datetime, Message] = Arg(key),
    ):
        if isinstance(input, datetime):
            return

        plaintext = input.extract_plain_text()
        try:
            state[key] = get_datetime_fromisoformat_with_timezone(plaintext)
        except ValueError:
            await matcher.reject_arg(key, "请输入正确的日期，不然我没法理解呢！")

    return _key_parser


@wordcloud_cmd.handle(parameterless=[Depends(ensure_group)])
async def handle_first_receive(
    state: T_State, type: Optional[str] = None, time: Optional[str] = None
):
    dt = get_datetime_now_with_timezone()

    if not type:
        await wordcloud_cmd.finish(__plugin_meta__.usage)

    if type == "今日":
        state["start"] = dt.replace(hour=0, minute=0, second=0, microsecond=0)
        state["stop"] = dt
    elif type == "昨日":
        state["stop"] = dt.replace(hour=0, minute=0, second=0, microsecond=0)
        state["start"] = state["stop"] - timedelta(days=1)
    elif type == "本周":
        state["start"] = dt.replace(
            hour=0, minute=0, second=0, microsecond=0
        ) - timedelta(days=dt.weekday())
        state["stop"] = dt
    elif type == "上周":
        state["stop"] = dt.replace(
            hour=0, minute=0, second=0, microsecond=0
        ) - timedelta(days=dt.weekday())
        state["start"] = state["stop"] - timedelta(days=7)
    elif type == "本月":
        state["start"] = dt.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
        state["stop"] = dt
    elif type == "上月":
        state["stop"] = dt.replace(
            day=1, hour=0, minute=0, second=0, microsecond=0
        ) - timedelta(microseconds=1)
        state["start"] = state["stop"].replace(
            day=1, hour=0, minute=0, second=0, microsecond=0
        )
    elif type == "年度":
        state["start"] = dt.replace(
            month=1, day=1, hour=0, minute=0, second=0, microsecond=0
        )
        state["stop"] = dt
    elif type == "历史":
        if time:
            plaintext = time
            if match := re.match(r"^(.+?)(?:~(.+))?$", plaintext):
                start = match[1]
                stop = match[2]
                try:
                    state["start"] = get_datetime_fromisoformat_with_timezone(start)
                    if stop:
                        state["stop"] = get_datetime_fromisoformat_with_timezone(stop)
                    else:
                        # 如果没有指定结束日期，则认为是所给日期的当天的词云
                        state["start"] = state["start"].replace(
                            hour=0, minute=0, second=0, microsecond=0
                        )
                        state["stop"] = state["start"] + timedelta(days=1)
                except ValueError:
                    await wordcloud_cmd.finish("请输入正确的日期，不然我没法理解呢！")


@wordcloud_cmd.got(
    "start",
    prompt="请输入你要查询的起始日期（如 2022-01-01）",
    parameterless=[Depends(parse_datetime("start"))],
)
@wordcloud_cmd.got(
    "stop",
    prompt="请输入你要查询的结束日期（如 2022-02-22）",
    parameterless=[Depends(parse_datetime("stop"))],
)
async def handle_wordcloud(
    my: Query[bool] = AlconnaQuery("my.value", False),
    session: Session = Depends(extract_session),
    start: datetime = Arg(),
    stop: datetime = Arg(),
    mask_key: str = Depends(get_mask_key),
):
    """生成词云"""
    messages = await get_messages_plain_text(
        session=session,
        id_type=SessionIdType.GROUP_USER if my.result else SessionIdType.GROUP,
        include_bot_id=False,
        include_bot_type=False,
        types=["message"],  # 排除机器人自己发的消息
        time_start=start,
        time_stop=stop,
        exclude_id1s=plugin_config.wordcloud_exclude_user_ids,
    )

    if not (image := await get_wordcloud(messages, mask_key)):
        await wordcloud_cmd.finish("没有足够的数据生成词云", at_sender=my.result)
        return  # pragma: no cover

    await saa.Image(image, "wordcloud.png").finish(at_sender=my.result)


set_mask_cmd = on_alconna(
    Alconna(
        "设置词云形状",
        Option("--default", default=False, action=store_true),
        Args["img?", alc.Image],
    ),
    permission=admin_permission(),
    use_cmd_start=True,
)
set_mask_cmd.shortcut(
    "设置词云默认形状",
    {
        "prefix": True,
        "command": "设置词云形状",
        "args": ["--default"],
    },
)


@set_mask_cmd.handle(parameterless=[Depends(ensure_group)])
async def _(
    matcher: AlconnaMatcher,
    img: Match[bytes] = AlconnaMatch("img", image_fetch),
):
    if img.available:
        matcher.set_path_arg("img", img.result)


@set_mask_cmd.got_path("img", "请发送一张图片作为词云形状", image_fetch)
async def handle_save_mask(
    bot: Bot,
    event: Event,
    img: bytes,
    default: Query[bool] = AlconnaQuery("default.value", default=False),
    mask_key: str = Depends(get_mask_key),
):
    mask = Image.open(BytesIO(img))
    if default.result:
        if not await SUPERUSER(bot, event):
            await set_mask_cmd.finish("仅超级用户可设置词云默认形状")
        mask.save(plugin_config.get_mask_path(), format="PNG")
        await set_mask_cmd.finish("词云默认形状设置成功")
    else:
        mask.save(plugin_config.get_mask_path(mask_key), format="PNG")
        await set_mask_cmd.finish("词云形状设置成功")


remove_mask_cmd = on_alconna(
    Alconna(
        "删除词云形状",
        Option("--default", default=False, action=store_true),
    ),
    permission=admin_permission(),
    use_cmd_start=True,
)
remove_mask_cmd.shortcut(
    "删除词云默认形状",
    {
        "prefix": True,
        "command": "删除词云形状",
        "args": ["--default"],
    },
)


@remove_mask_cmd.handle(parameterless=[Depends(ensure_group)])
async def _(
    bot: Bot,
    event: Event,
    default: Query[bool] = AlconnaQuery("default.value", default=False),
    mask_key: str = Depends(get_mask_key),
):
    if default.result:
        if not await SUPERUSER(bot, event):
            await remove_mask_cmd.finish("仅超级用户可删除词云默认形状")
        mask_path = plugin_config.get_mask_path()
        mask_path.unlink(missing_ok=True)
        await remove_mask_cmd.finish("词云默认形状已删除")
    else:
        mask_path = plugin_config.get_mask_path(mask_key)
        mask_path.unlink(missing_ok=True)
        await remove_mask_cmd.finish("词云形状已删除")


schedule_cmd = on_alconna(
    Alconna(
        "词云定时发送",
        Option("--action", Args["action_type", ["状态", "开启", "关闭"]], default="状态"),
        Args["type", ["每日"]]["time?", str],
    ),
    permission=admin_permission(),
    use_cmd_start=True,
)
schedule_cmd.shortcut(
    r"词云(?P<type>.+)定时发送状态",
    {
        "prefix": True,
        "command": "词云定时发送",
        "args": ["--action", "状态", "{type}"],
    },
)
schedule_cmd.shortcut(
    r"(?P<action>.+)词云(?P<type>.+)定时发送",
    {
        "prefix": True,
        "command": "词云定时发送",
        "args": ["--action", "{action}", "{type}"],
    },
)


@schedule_cmd.handle(parameterless=[Depends(ensure_group)])
async def _(
    time: Optional[str] = None,
    action_type: Query[str] = AlconnaQuery("action.action_type.value", "状态"),
    target: saa.PlatformTarget = Depends(saa.get_target),
):
    if action_type.result == "状态":
        schedule_time = await schedule_service.get_schedule(target)
        await schedule_cmd.finish(
            f"词云每日定时发送已开启，发送时间为：{schedule_time}" if schedule_time else "词云每日定时发送未开启"
        )
    elif action_type.result == "开启":
        schedule_time = None
        if time:
            try:
                schedule_time = get_time_fromisoformat_with_timezone(time)
            except ValueError:
                await schedule_cmd.finish("请输入正确的时间，不然我没法理解呢！")
        await schedule_service.add_schedule(target, time=schedule_time)
        await schedule_cmd.finish(
            f"已开启词云每日定时发送，发送时间为：{schedule_time}"
            if schedule_time
            else f"已开启词云每日定时发送，发送时间为：{plugin_config.wordcloud_default_schedule_time}"
        )
    elif action_type.result == "关闭":
        await schedule_service.remove_schedule(target)
        await schedule_cmd.finish("已关闭词云每日定时发送")
