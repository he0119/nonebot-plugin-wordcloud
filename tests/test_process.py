def process_messages():
    from nonebot_plugin_wordcloud.data_source import analyse_messages

    msgs = [
        "这是一个奇怪的句子。",
        "1😅🟨二",
        "1👨🏿‍🔧👨‍👨‍👧‍👧🛀🏽🧑🏽‍❤️‍🧑🏾1",
        "1  2",
        "1 http://v2.nonebot.dev/ 2",
        "1 https://v2.nonebot.dev/ 2",
        "1 https://api.weibo.cn/share/312975272,47087.html?weibo_id=4770873388 2",
    ]
    frequencies = analyse_messages(msgs)

    # You may return anything you want, like the result of a computation
    return frequencies


def test_process_messages(benchmark, snapshot):
    frequencies = benchmark(process_messages)

    assert frequencies == snapshot
