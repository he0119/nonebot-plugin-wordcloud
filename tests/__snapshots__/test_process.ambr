# serializer version: 1
# name: test_process_messages
  dict({
    '11': 2.39095350058,
    '一个': 0.5635101944259999,
    '句子': 1.8226371818379998,
    '奇怪': 1.292106194262,
    '这是': 0.858325655278,
  })
# ---
